dashboard/test/controllers/levels_controller_test.rb

```diff
@@ -13,6 +13,13 @@ class LevelsControllerTest < ActionController::TestCase
     @levelbuilder = create(:levelbuilder)
     sign_in(@levelbuilder)
     @program = '<hey/>'
+
+    # Ensure that we get a real S3 url for the level source image when
+    # requested, and that we never try to upload an image to S3.
+
+    CDO.stubs(:disable_s3_image_uploads).returns(false)
+    LevelSourceImage.any_instance.expects(:save_to_s3).never
+    create(:level_source, :with_image, level: @level, data: @program)
```

Comment:
The preceding comment relates to these 3 lines. I was trying to only [tell you why](https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/), but maybe I took it too far. I've extracted a method for this and documented the steps more clearly.